## AOC 2024/11

### Part 1

--- Day 11: Plutonian Pebbles ---

The ancient civilization on Pluto was known for its ability to manipulate spacetime, and while The Historians explore their infinite corridors, you've noticed a strange set of physics-defying stones.

At first glance, they seem like normal stones: they're arranged in a perfectly straight line, and each stone has a number engraved on it.

The strange part is that every time you blink, the stones change.

Sometimes, the number engraved on a stone changes. Other times, a stone might split in two, causing all the other stones to shift over a bit to make room in their perfectly straight line.

As you observe them for a while, you find that the stones have a consistent behavior. Every time you blink, the stones each simultaneously change according to the first applicable rule in this list:

- If the stone is engraved with the number 0, it is replaced by a stone engraved with the number 1.
- If the stone is engraved with a number that has an even number of digits, it is replaced by two stones. The left half of the digits are engraved on the new left stone, and the right half of the digits are engraved on the new right stone. (The new numbers don't keep extra leading zeroes: 1000 would become stones 10 and 0.)
- If none of the other rules apply, the stone is replaced by a new stone; the old stone's number multiplied by 2024 is engraved on the new stone.
- No matter how the stones change, their order is preserved, and they stay on their perfectly straight line.

How will the stones evolve if you keep blinking at them? You take a note of the number engraved on each stone in the line (your puzzle input).

If you have an arrangement of five stones engraved with the numbers 0 1 10 99 999 and you blink once, the stones transform as follows:

- The first stone, 0, becomes a stone marked 1.
- The second stone, 1, is multiplied by 2024 to become 2024.
- The third stone, 10, is split into a stone marked 1 followed by a stone marked 0.
- The fourth stone, 99, is split into two stones marked 9.
- The fifth stone, 999, is replaced by a stone marked 2021976.
So, after blinking once, your five stones would become an arrangement of seven stones engraved with the numbers 1 2024 1 0 9 9 2021976.

Here is a longer example:

- Initial arrangement: `125 17`
- After 1 blink: `253000 1 7`
- After 2 blinks: `253 0 2024 14168`
- After 3 blinks: `512072 1 20 24 28676032`
- After 4 blinks: `512 72 2024 2 0 2 4 2867 6032`
- After 5 blinks: `1036288 7 2 20 24 4048 1 4048 8096 28 67 60 32`
- After 6 blinks: `2097446912 14168 4048 2 0 2 4 40 48 2024 40 48 80 96 2 8 6 7 6 0 3 2`

In this example, after blinking six times, you would have 22 stones. After blinking 25 times, you would have 55312 stones!

Consider the arrangement of stones in front of you. How many stones will you have after blinking 25 times?

**AI Prompt**

I'm a skilled python developer, please help go consider various approaches for solving this problem and their pros/cons. I prefer simpler and understandable solutions, easily readable also for other developers. It's ok to generate code with testcases without explanation. I'll ask if I do not understand it. Always wrap test code into a function, to not clutter the global variable namespace.

**AI Response**

For this problem, we have several possible approaches. Let's consider the main ones:

1. Direct List Simulation
- Pros: Straightforward, easy to understand and debug
- Cons: Could be memory-intensive for large numbers of blinks
- Uses a list to store numbers and applies transformations directly

2. Generator/Iterator Approach
- Pros: Memory efficient, follows functional programming principles
- Cons: Slightly more complex to implement
- Yields new states without storing full history

Let me start with a test case implementation to validate our basic rules:

```python
def test_stone_rules():
    def apply_rules(nums):
        result = []
        for n in nums:
            if n == 0:
                result.append(1)
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                left = int(s[:mid])
                right = int(s[mid:])
                result.extend([left, right])
            else:
                result.append(n * 2024)
        return result
    
    # Test case 1
    assert apply_rules([0, 1, 10, 99, 999]) == [1, 2024, 1, 0, 9, 9, 2021976]
    
    # Test case 2
    assert apply_rules([125, 17]) == [253000, 1, 7]
    
    return "All tests passed!"

print(test_stone_rules())
```

Would you like me to implement one of these approaches, or would you prefer to explore a different solution strategy?

In [ ]:
def test_stone_rules():
    def apply_rules(nums):
        result = []
        for n in nums:
            if n == 0:
                result.append(1)
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                left = int(s[:mid])
                right = int(s[mid:])
                result.extend([left, right])
            else:
                result.append(n * 2024)
        return result
    
    # Test case 1
    assert apply_rules([0, 1, 10, 99, 999]) == [1, 2024, 1, 0, 9, 9, 2021976]
    
    # Test case 2
    assert apply_rules([125, 17]) == [253000, 1, 7]
    
    return "All tests passed!"

print(test_stone_rules())

All tests passed!


In [ ]:
sample_input = [125, 17]

**AI Prompt**

lets do the generator approach

**AI Response**

I'll help you create a generator-based solution. Here's a basic implementation:

```python
def stone_generator(initial_stones, num_blinks):
    current = initial_stones
    
    for _ in range(num_blinks):
        next_state = []
        for n in current:
            if n == 0:
                next_state.append(1)
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                next_state.extend([int(s[:mid]), int(s[mid:])])
            else:
                next_state.append(n * 2024)
        current = next_state
        yield len(current)

def solve_part1(stones):
    return list(stone_generator(stones, 25))[-1]

# Test
def test_solution():
    test_input = [125, 17]
    steps = list(stone_generator(test_input, 6))
    assert steps[0] == 3  # After 1 blink
    assert steps[5] == 22  # After 6 blinks
    print("Tests passed!")

test_solution()
```

This generator yields the length at each step rather than the full state, since we only need the final count. Would you like me to modify it to track the full states instead, or shall we proceed with implementing the full solution?

In [ ]:
def stone_generator(initial_stones, num_blinks):
    current = initial_stones
    
    for _ in range(num_blinks):
        next_state = []
        for n in current:
            if n == 0:
                next_state.append(1)
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                next_state.extend([int(s[:mid]), int(s[mid:])])
            else:
                next_state.append(n * 2024)
        current = next_state
        yield len(current)

def solve_part1(stones):
    return list(stone_generator(stones, 25))[-1]

# Test
def test_solution():
    steps = list(stone_generator(sample_input, 25))
    assert steps[0] == 3  # After 1 blink
    assert steps[5] == 22  # After 6 blinks
    assert steps[24] == 55312   # After 25 blinks
    print("Tests passed!")

test_solution()

Tests passed!


**AI Prompt**

get full aoc data

**AI Response**

I'll help you get the AOC data using the `aocd` library. Here's how we can modify the solution to use real data:

```python
from aocd import get_data

def get_input():
    data = get_data(day=11, year=2024)
    return [int(x) for x in data.split()]

def solve():
    stones = get_input()
    result = solve_part1(stones)
    print(f"Part 1: {result}")

solve()
```

Would you like to run this? Note that you'll need to have your AOC session token configured for `aocd` to work. Let me know if you need help with that setup.

In [ ]:
from aocd import get_data
full_input = [int(s) for s in get_data(day=11, year=2024).split()]
full_input

[5178527, 8525, 22, 376299, 3, 69312, 0, 275]

In [ ]:
def solve_part1(N: int):
    steps = list(stone_generator(full_input, N))
    print(steps)
    print(f"Num stones {steps[N-1]}")

solve_part1(25)

[11, 15, 24, 34, 44, 68, 119, 160, 229, 359, 553, 848, 1242, 1920, 2922, 4257, 6733, 10265, 15184, 23212, 35180, 54399, 81708, 122561, 189547]
Num stones 189547


the answer was correct, now lets solve part2

### Part2

The Historians sure are taking a long time. To be fair, the infinite corridors are very large.

How many stones would you have after blinking a total of 75 times?

**AI Prompt**

what is the time/space complexity of the `stone_generator`?

**AI Response**

Let's analyze the complexity:

Time Complexity:
- For each blink, we process every stone
- The number of stones grows with each blink
- Looking at our output, we can see the growth is significant:
  - Started with 8 stones
  - After 25 blinks: 189,547 stones
  - Each stone requires O(1) to O(log n) operations (string conversion and splitting)

So for blink k, if we have n(k) stones:
- Time: O(n(k)) to O(n(k) * log(max_number))
- Space: O(n(k)) for the next_state list

The sequence appears to grow exponentially, making it impractical for 75 blinks with the current approach. Would you like to analyze the growth pattern to see if we can find a mathematical pattern or optimization?

**AI Prompt**

what optimisations could we implement?

**AI Response**

Let's consider possible optimizations:

1. Pattern Detection
- The sequence might have a repeating pattern or growth rate
- We could track the growth rate between steps to see if it's predictable
- If we find a pattern, we could skip computation and calculate directly

2. State Compression
- Instead of storing each number individually, we could store counts of identical numbers
- Format: (number, count) pairs
- This would help if many identical numbers appear

3. Mathematical Analysis
- Each operation (×2024, splitting, 0→1) follows fixed rules
- We might be able to predict how many numbers will split vs multiply in each generation
- Could lead to a mathematical formula for growth

Would you like to explore any of these approaches? We could start by analyzing the growth pattern or implementing state compression to see which might be most promising.

**AI Prompt**

would it suffice to keep just counts of different stone types?

**AI Response**

Yes, that's a clever optimization! Instead of tracking individual stones, we could track counts of stones that behave the same way. 

For any given number:
1. All 0s become 1s
2. Numbers with even digits split in a deterministic way
3. Other numbers just multiply by 2024

So we could maintain a dictionary/counter where:
- key: the number
- value: how many times this number appears

Would you like me to sketch out how this modified approach would work? It should dramatically reduce both time and space complexity since many stones will be duplicates.

**AI Prompt**

please refactor `stone_generator2` from `stone_generator` using this counts approach. Write testcases to compare values returned from `stone_generator2` and `stone_generator`

**AI Response**

Here's the implementation with test cases:

```python
from collections import Counter

def stone_generator2(initial_stones, num_blinks):
    current = Counter(initial_stones)
    
    for _ in range(num_blinks):
        next_state = Counter()
        for n, count in current.items():
            if n == 0:
                next_state[1] += count
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                left = int(s[:mid])
                right = int(s[mid:])
                next_state[left] += count
                next_state[right] += count
            else:
                next_state[n * 2024] += count
        current = next_state
        yield sum(current.values())

def test_generators():
    test_cases = [
        [125, 17],  # original test case
        full_input, # actual input
    ]
    
    for test_input in test_cases:
        gen1 = list(stone_generator(test_input, 25))
        gen2 = list(stone_generator2(test_input, 25))
        assert gen1 == gen2, f"Mismatch: {gen1} != {gen2}"
    print("All generator tests passed!")

test_generators()
```

Would you like to try the optimized version with 75 blinks now?

In [ ]:
from collections import Counter

def stone_generator2(initial_stones, num_blinks):
    current = Counter(initial_stones)
    
    for _ in range(num_blinks):
        next_state = Counter()
        for n, count in current.items():
            if n == 0:
                next_state[1] += count
            elif len(str(n)) % 2 == 0:
                s = str(n)
                mid = len(s) // 2
                left = int(s[:mid])
                right = int(s[mid:])
                next_state[left] += count
                next_state[right] += count
            else:
                next_state[n * 2024] += count
        current = next_state
        yield sum(current.values())

def test_generators():
    test_cases = [
        [125, 17],  # original test case
        full_input, # actual input
    ]
    
    for test_input in test_cases:
        gen1 = list(stone_generator(test_input, 25))
        gen2 = list(stone_generator2(test_input, 25))
        assert gen1 == gen2, f"Mismatch: {gen1} != {gen2}"
    print("All generator tests passed!")

test_generators()

All generator tests passed!


In [ ]:
def solve_part2(N: int):
    steps = list(stone_generator2(full_input, N))
    print(steps)
    print(f"Num stones {steps[N-1]}")

solve_part2(75)

[11, 15, 24, 34, 44, 68, 119, 160, 229, 359, 553, 848, 1242, 1920, 2922, 4257, 6733, 10265, 15184, 23212, 35180, 54399, 81708, 122561, 189547, 285429, 433672, 661693, 998273, 1525214, 2306931, 3508534, 5351940, 8067746, 12309881, 18705685, 28336072, 43176239, 65378277, 99491570, 151182056, 229081370, 348989228, 529126588, 803609759, 1222382207, 1853535428, 2819248327, 4279506546, 6497541585, 9881561849, 14989831626, 22783880272, 34611747294, 52535276048, 79865768259, 121234753538, 184178398765, 279837397329, 424797840901, 645605535731, 980372352184, 1488934975775, 2262406077367, 3435039839919, 5218963206220, 7926948856006, 12038235032256, 18290482441458, 27775066436548, 42192151976831, 64091881828909, 97332869199706, 147869130284377, 224577979481346]
Num stones 224577979481346
